pallets/allocations/src/lib.rs

```diff
@@ -82,57 +81,83 @@ impl<T: Config> MintCurve<T> {
 			maximum_supply,
 		}
 	}
 
-	pub fn checked_calc_next_session_quota(
+	pub fn calc_session_quota(
 		&self,
-		block_number: T::BlockNumber,
+		n: T::BlockNumber,
+		curve_start: T::BlockNumber,
 		current_supply: BalanceOf<T>,
-		forced: bool,
-	) -> Option<BalanceOf<T>> {
-		if (block_number.checked_rem(&self.fiscal_period) == Some(T::BlockNumber::zero())) || forced {
-			let step: usize = block_number
-				.checked_div(&self.fiscal_period)
-				.unwrap_or_else(Zero::zero)
-				.unique_saturated_into();
-			let max_inflation_rate = *self
-				.inflation_steps
-				.get(step)
-				.or_else(|| self.inflation_steps.last())
-				.unwrap_or(&Zero::zero());
-			let target_increase =
-				(self.maximum_supply.saturating_sub(current_supply)).min(max_inflation_rate * current_supply);
-			let session_quota = Perbill::from_rational(self.session_period, self.fiscal_period) * target_increase;
-			Some(session_quota)
-		} else {
-			None
-		}
+	) -> BalanceOf<T> {
+		let step: usize = n
+			.saturating_sub(curve_start)
+			.checked_div(&self.fiscal_period)
+			.unwrap_or_else(Bounded::max_value)
+			.unique_saturated_into();
+		let max_inflation_rate = *self
+			.inflation_steps
+			.get(step)
+			.or_else(|| self.inflation_steps.last())
+			.unwrap_or(&Zero::zero());
+		let target_increase =
+			(self.maximum_supply.saturating_sub(current_supply)).min(max_inflation_rate * current_supply);
+		Perbill::from_rational(self.session_period, self.fiscal_period) * target_increase
 	}
 
-	pub fn should_update_session_quota(&self, block_number: T::BlockNumber) -> bool {
-		block_number.checked_rem(&self.session_period) == Some(T::BlockNumber::zero())
+	pub fn next_quota_renew_schedule(&self, n: T::BlockNumber, curve_start: T::BlockNumber) -> T::BlockNumber {
+		Self::next_schedule(n, curve_start, self.session_period)
+	}
+
+	pub fn next_quota_calc_schedule(&self, n: T::BlockNumber, curve_start: T::BlockNumber) -> T::BlockNumber {
+		Self::next_schedule(n, curve_start, self.fiscal_period)
 	}
 
 	#[inline(always)]
 	pub fn session_period(&self) -> T::BlockNumber {
 		self.session_period
 	}
+
 	#[inline(always)]
 	pub fn fiscal_period(&self) -> T::BlockNumber {
 		self.fiscal_period
 	}
+
 	#[inline(always)]
 	pub fn maximum_supply(&self) -> BalanceOf<T> {
 		self.maximum_supply
 	}
+
+	/// Helper function to calculate the very next schedule based on the current block number.
+	fn next_schedule(n: T::BlockNumber, curve_start: T::BlockNumber, period: T::BlockNumber) -> T::BlockNumber {
+		if n >= curve_start {
+			n.saturating_sub(curve_start)
+				.checked_div(&period)
+				.unwrap_or_else(Bounded::max_value)
+				.saturating_add(One::one())
+				.saturating_mul(period)
+				.saturating_add(curve_start)
+		} else {
+			let s = curve_start.saturating_sub(
+				curve_start
+					.saturating_sub(n)
+					.checked_div(&period)
+					.unwrap_or_else(Bounded::max_value)
+					.saturating_mul(period),
+			);
+			if n == s {
+				n.saturating_add(period)
```

Comment:
will this not drift if call is delayed?

Reply:
An extrinsic is executed in an atomic way, meaning there wouldn't be a drift in the block number or an execution of any other extrinsic in parallel. This should be through for the relay chain block number as well, because every single call should be executed by parachain validators to be backed in a certain relay chain block.